cdc/puller/multiplexing_puller.go

```diff
@@ -0,0 +1,428 @@
+	p.unsubscribe(spans)
+}
+
+func (p *MultiplexingPuller) unsubscribe(spans []tablepb.Span) {
+	for _, span := range spans {
+		if subID, ok := p.client.Unsubscribe(span); ok {
+			progress := p.subscriptions.m[subID]
+			progress.consume.Lock()
+			progress.consume.removed = true
+			progress.consume.Unlock()
+
+			delete(p.subscriptions.m, subID)
+			p.subscriptions.n.Delete(span)
+			continue
+		}
+		log.Panic("unexist unsubscription",
+			zap.String("namespace", p.changefeed.Namespace),
+			zap.String("changefeed", p.changefeed.ID),
+			zap.String("span", span.String()))
+	}
+}
+
+// Run the puller.
+func (p *MultiplexingPuller) Run(ctx context.Context) (err error) {
+	return p.run(ctx, true)
+}
+
+func (p *MultiplexingPuller) run(ctx context.Context, includeClient bool) error {
+	p.pullerEventCounterKv = PullerEventCounter.WithLabelValues(p.changefeed.Namespace, p.changefeed.ID, "kv")
+	p.pullerEventCounterResolved = PullerEventCounter.WithLabelValues(p.changefeed.Namespace, p.changefeed.ID, "resolved")
+	p.queueKvDuration = pullerQueueDuration.WithLabelValues(p.changefeed.Namespace, p.changefeed.ID, "kv")
+	p.queueResolvedDuration = pullerQueueDuration.WithLabelValues(p.changefeed.Namespace, p.changefeed.ID, "resolved")
+	defer func() {
+		PullerEventCounter.DeleteLabelValues(p.changefeed.Namespace, p.changefeed.ID, "kv")
+		PullerEventCounter.DeleteLabelValues(p.changefeed.Namespace, p.changefeed.ID, "resolved")
+		pullerQueueDuration.DeleteLabelValues(p.changefeed.Namespace, p.changefeed.ID, "kv")
+		pullerQueueDuration.DeleteLabelValues(p.changefeed.Namespace, p.changefeed.ID, "resolved")
+	}()
+
+	g, ctx := errgroup.WithContext(ctx)
+	if includeClient {
+		g.Go(func() error { return p.client.Run(ctx) })
+	}
+
+	g.Go(func() error { return p.checkResolveLock(ctx) })
+
+	for i := 0; i < p.frontiers; i++ {
+		g.Go(func() error { return p.advanceSpans(ctx) })
+	}
+	for i := range p.inputChs {
+		inputCh := p.inputChs[i]
+		g.Go(func() error { return p.handleInputCh(ctx, inputCh) })
+	}
+
+	log.Info("MultiplexingPuller starts",
+		zap.String("namespace", p.changefeed.Namespace),
+		zap.String("changefeed", p.changefeed.ID),
+		zap.Int("frontierConcurrent", p.frontiers))
+	return g.Wait()
+}
+
+func (p *MultiplexingPuller) handleInputCh(ctx context.Context, inputCh <-chan kv.MultiplexingEvent) error {
+	for {
+		var e kv.MultiplexingEvent
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case e = <-inputCh:
+		}
+
+		progress := p.getProgress(e.SubscriptionID)
+		if progress == nil {
+			continue
+		}
+
+		if e.Val != nil {
+			p.queueKvDuration.Observe(float64(time.Since(e.Start).Milliseconds()))
+			p.pullerEventCounterKv.Inc()
+			if err := progress.consume.f(ctx, e.Val, progress.spans); err != nil {
+				return errors.Trace(err)
+			}
+		} else if e.Resolved != nil {
+			p.pullerEventCounterResolved.Add(float64(len(e.Resolved.Spans)))
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case progress.resolvedSpans <- e:
```

Comment:
Is it possible for the worker to get stuck here?

Reply:
I guess no. frontier threads just do some CPU intensive works. If we move those works into this goroutine, it will also stuck.